aaronthangnguyen/cen4010-swe1
# Add `get books by genre`, `get books by top sellers`, `apply discount by publisher` endpoints

### Summary

We've elevated our book management system by introducing two pivotal `GET` endpoints. The first allows users to retrieve a curated list of books based on their genre, streamlining the search process and catering to specific reading preferences. The second showcases our top 10 best-selling books, offering a quick insight into the titles that have resonated most with our audience based on copies sold. These enhancements optimize the browsing experience by providing tailored book recommendations and highlighting trending literature, ensuring our readers always have access to the best of our catalog.

### Changes

- BookController: Implemented the `GET` endpoint to retrieve books by genre and top-selling books.
- BookService: Added two service methods to retrieve books by genre and top-selling books.
- BookRepository: Update the necessary query for fetching books by genre and top-selling books.

### Endpoint Details
- Purpose: Allows users or administrators to retrieve a list of books by genre and top-selling books.
- Endpoint 1: `/api/books/genre/{genre}`
- Endpoint 2: `/api/books/top-seller`
- HTTP Method: `GET`
- Response 1: Returns a JSON array containing detailed information about all books by genre in the system. Each book entry includes essential details such as the book's ISBN, title, description, price, author, genre, publisher, publication year, the number of copies sold, and rating (if have).
- Response 2: Returns a JSON array containing detailed information about the top 10 selling books in the system. Each book entry includes essential details such as the book's ISBN, title, description, price, author, genre, publisher, publication year, the number of copies sold, and rating (if have).
    Example Response:
```json
[
 {
  "id":"948e545e-b7bb-4594-8f8f-52259167e416",
  "isbn":"9780439064873",
  "name":"Harry Potter and the Chamber of Secrets",
  "description":"The second book in the Harry Potter series.",
  "price":20.99,"authorId":"c88664b2-2adb-4d1a-bc6c-0a2916be3e8c",
  "genre":"Fantasy",
  "publisher":"Scholastic",
  "yearPublished":1998,
  "copiesSold":77000000
 }
]
```

Test Cases
- Retrieve the list of all books by genre / by top-seller.
- Check for an empty list if there are no books in the system.
- Verify the response structure and data accuracy.

How to Test
- Pull the branch to your local setup.
- Run the application.
- Use the provided curl command or any API tool like Postman to test the endpoint.

```cmd
curl http://localhost:8080/api/books/genre/fantasy
curl http://localhost:8080/api/books/top-seller
```

### Update: Book Discount by publisher
We want our book system to have a good deal with customers, so we want to use the `PUT` or `PATCH` Request Type to update the price by the percentage of discount and publisher. This enhancement help the customer to have a better deal with publisher that have sale or discount deals, ensuring the customer always get what they want. 

### Changes

- Book: If price under $0.00, return error.
- BookController: Implemented the `PATCH` endpoint to modify the books price based on discount percentage and publisher.
- BookService: Added one service to update price based on publisher.
- BookRepository: Added one neccesary queue to returnthe publisher.
- DiscountDTO: Added class for discount constructor.
- PublisherNotFoundException: return error if publisher not found. 

### Endpoint details
- Purpose: Allows administrators to update the price of all books under a publisher by a discount percentage.
- Endpoint: `/api/books/discount`
- HTTP Method: `PATCH`
- Parameters: Discount Percent, Publisher
- Response: Returns a string message saying all the books by publisher discount by *% and updates the database with the discount percent with the publisher.
- Example:
  Input: ```curl -X PATCH http://localhost:8080/api/books/discount -H "Content-Type: application/json" -d "{\"discountPercent\": 10, \"publisher\": \"Example Publisher\"}"```
  Response:
(Response can be none return but this is to make sure we have good output)
All books by publisher 'Example Publisher' discounted by 10.0%.

Test Cases
- The discount percentage is applied to books within the publisher.
- Check for the publisher if it does have books. 
- Verify the response structure and data accuracy.

How to Test
- Pull the branch to your local setup.
- Run the application.
- Use the provided curl command or any API tool like Postman to test the endpoint.
